### ISSUE TITLE
[Bug] Mouse emulation lacks arithmetic saturation [sf#498]

### ISSUE THREAD

**Issue by andete** [openMSX/openMSX#812]
**Reported by sd-snatcher on 2013-08-31 17:03 UTC**
This bug is related to the #477.

Interesting enough, just like what happened on Trackball.cc, the mouse emulation also fails arithmetic saturation tests. This means that the same overflow bug happens with the mouse, albeit they're much rarer because the offsets are 8-bit.

But it seems that there was already an attempt to implement the arithmetic saturation on Mouse.cc, but the code is bugged and end up imposing limits to the actual X/Y coordinates instead of the offsets.

The fixed code for the lines 240-243 of Mouse.cc seems to be:

int newX = curxrel - mev.getX();
int newY = curyrel - mev.getY();
int deltaX = max(127, min(-128, newX - curxrel));
int deltaY = max(127, min(-128, newY - curyrel));

**Comment by andete** [comment]
**Updated by m9710797 on 2013-09-01 19:00 UTC**
- **status**: open --> closed-fixed

**Comment by andete** [comment]
**Commented by m9710797 on 2013-09-01 19:00 UTC**
You're right that the original clipping code in the mouse emulation didn't work (it actually clipped on -256..254 instead of -128..127). But I just tested a real MSX mouse (Philips SBC3810) and it doesn't do any clipping at all (so by accident the buggy emulation code was already (closer to) the real behavior).

In revision ebeb2989 I've completely removed the mouse-offset clipping code.

**Comment by andete** [comment]
**Commented by sd-snatcher on 2013-09-02 02:16 UTC**

The Philips mouse may not have clipping, but I have a MK-Mouse II here and it clearly has.

The clipping seems to be implementation dependent. Some mice have, others don't. This happens because the MSX mice are implemented by software on a small microcontroller, so each mouse has a unique algorithm.

If you want to, I may record a video of my MK-Mouse II, similar to the one Werner recorded for the trackball.

**Comment by andete** [comment]
**Commented by m9710797 on 2013-09-02 08:10 UTC**
You can make a video, but I don't think it's really needed (the trackball video _was_ very useful). This is how I tested:

  10 A=PAD(12)
  20 X=PAD(13)
  30 Y=PAD(14)
  40 PRINT X,Y
  50 FORI=0to250;NEXT
  60 GOTO 10

By varying the value '250' in line 50 it's possible to control how frequently the mouse is sampled. I preferred this over physically moving the mouse very fast because I also want to make sure there is no 'mechanical overflow' (I mean if the led/photo-diode doesn't sample the holes fast enough (or however the mechanical construction works) then you'll get aliasing effects and the axis may appear to rotate in the opposite direction).

Anyway, to get 8-bit overflow you already need to move the mouse over (more than) half
the screen width within one mouse-poll-interval (typically 1 or a few 1/50s). That's very fast. I don't believe this is worth the complexity of supporting different mouse models.

**Comment by andete** [comment]
**Commented by sd-snatcher on 2013-09-03 13:37 UTC**

Excellent idea! This test is at the same time simple and very well thought.

I tested the MK-Mouse II using it, and it indeed overflows this way. Then I opened the mouse to check why it didn't overflow in the fast test.

What happens is that the mechanical parts indeed act as a restriction. They just can't spin fast enough to make the counters overflow. The ball will loose traction over the surface, and will also loose traction over the internal wheels. 

In a nutshell, the design has a RPM limit, and that itself act as a "mechanical clip" for the offsets.
